apps/glusterfs/operations_block_restriction.go

```diff
@@ -0,0 +1,164 @@
+// that change the block restriction state of a volume.
+type VolumeSetBlockRestrictionOperation struct {
+	OperationManager
+	noRetriesOperation
+	vol         *VolumeEntry
+	restriction api.BlockRestriction
+}
+
+// NewVolumeSetBlockRestrictionOperation returns a new
+// VolumeSetBlockRestrictionOperation populated with the given params.
+func NewVolumeSetBlockRestrictionOperation(
+	vol *VolumeEntry, db wdb.DB,
+	restriction api.BlockRestriction) *VolumeSetBlockRestrictionOperation {
+
+	return &VolumeSetBlockRestrictionOperation{
+		OperationManager: OperationManager{
+			db: db,
+			op: NewPendingOperationEntry(NEW_ID),
+		},
+		vol:         vol,
+		restriction: restriction,
+	}
+}
+
+func (ro *VolumeSetBlockRestrictionOperation) Label() string {
+	return "Set Volume's Block Volumes Restriction State"
+}
+
+func (ro *VolumeSetBlockRestrictionOperation) ResourceUrl() string {
+	return fmt.Sprintf("/volumes/%v", ro.vol.Info.Id)
+}
+
+// Build sets the state when adding restrictions.
+func (ro *VolumeSetBlockRestrictionOperation) Build() error {
+	if !ro.vol.Info.Block {
+		return fmt.Errorf(
+			"Block restrictions can only be set on block hosting volumes")
+	}
+	// do a "pre flight check" of unlock stuff
+	if e := ro.checkCanUnlock(ro.vol, ro.restriction); e != nil {
+		return e
+	}
+	if ro.restriction == api.Unrestricted {
+		// relaxing the restriction on a volume is performed
+		// only in finalize, after any sanity checks are done.
+		return nil
+	}
+	return ro.db.Update(func(tx *bolt.Tx) error {
+		v, err := NewVolumeEntryFromId(tx, ro.vol.Info.Id)
+		if err != nil {
+			return err
+		}
+		v.Info.BlockInfo.Restriction = ro.restriction
+		return v.Save(tx)
+	})
+}
+
+// Exec creates new bricks and volume on the underlying glusterfs storage system.
+func (ro *VolumeSetBlockRestrictionOperation) Exec(executor executors.Executor) error {
+	// currently does nothing. should do gluster-block sanity checks in the
+	// future
+	return nil
+}
+
+// Finalize marks our new volume and brick db entries as no longer pending.
+func (ro *VolumeSetBlockRestrictionOperation) Finalize() error {
+	if ro.restriction != api.Unrestricted {
+		// change was already made by build
+		return nil
+	}
+	return ro.db.Update(func(tx *bolt.Tx) error {
+		v, err := NewVolumeEntryFromId(tx, ro.vol.Info.Id)
+		if err != nil {
+			return err
+		}
+		if e := ro.checkCanUnlock(v, ro.restriction); e != nil {
+			return e
+		}
+		if e := ro.fixReservedSize(v); e != nil {
+			return e
+		}
+		v.Info.BlockInfo.Restriction = ro.restriction
+		return v.Save(tx)
+	})
+}
+
+// Rollback does nothing for this operation type.
+func (ro *VolumeSetBlockRestrictionOperation) Rollback(executor executors.Executor) error {
+	return nil
+}
+
+func (ro *VolumeSetBlockRestrictionOperation) checkCanUnlock(
+	v *VolumeEntry, r api.BlockRestriction) error {
+
+	switch {
+	case v.Info.BlockInfo.Restriction == r:
+		// no change is always OK
```

Comment:
I like reading switch states where cases of are of same type. Hence, if we could have all current states handles first it would be nice. Like:
a. current state is locked and request is for...
b. current state is "" and request is for...
c. current state is "locked-by-update" and request is for...

Whatever checks remain apart from the above ones can come in later in the validation.

Reply:
Updated.